File: crack-data-structures-and-algorithms/leetcode/python-impl/swap_nodes_in_pair_q24.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

# 核心思想
# 用 extract-then-insert 调整一个pair，然后前进
class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        dummy = ListNode(0)
        dummy.next = head
        pp = dummy
        p = head
        # even length, p would finally point to nil
        # odd length, p would finally point to the last node.
        while p and p.next:
            pn = p.next
            p.next = pn.next
            pn.next = pp.next
            pp.next = pn
            # calibrate pp & p
            pp = p
            p = p.next

        return dummy.next

File: crack-data-structures-and-algorithms/leetcode/swap_nodes_in_pair_q24.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 核心思想
// 用 extract-then-insert 调整一个pair；调整后 p 指向的结点已经在后面，所以 p.next 就是下一个要调整的 pair 的第一个节点
// 终止条件是 p && p.next 有效
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head) {
            return nullptr;
        }

        ListNode dummy(0);
        dummy.next = head;

        auto pp = &dummy;
        auto p = head;
        while (p && p->next) {
            auto pn = p->next;
            p->next = pn->next;
            pn->next = pp->next;
            pp->next = pn;

            pp = p;
            p = p->next;
        }

        return dummy.next;
    }
};
